collectors/delta_distribution.go

```diff
@@ -0,0 +1,170 @@
+
+	// ListMetrics will return all known entries in the store for a metricDescriptorName
+	ListMetrics(metricDescriptorName string) map[string][]*CollectedHistogram
+}
+
+type histogramEntry = map[uint64]*CollectedHistogram
+
+type inMemoryDeltaDistributionStore struct {
+	store      map[string]histogramEntry
+	ttl        time.Duration
+	storeMutex *sync.RWMutex
+	logger     log.Logger
+}
+
+// NewInMemoryDeltaDistributionStore returns an implementation of DeltaDistributionStore which is persisted in-memory
+func NewInMemoryDeltaDistributionStore(logger log.Logger, ttl time.Duration) DeltaDistributionStore {
+	return inMemoryDeltaDistributionStore{
+		store:      map[string]histogramEntry{},
+		storeMutex: &sync.RWMutex{},
+		logger:     logger,
+		ttl:        ttl,
+	}
+}
+
+func (s inMemoryDeltaDistributionStore) Increment(metricDescriptor *monitoring.MetricDescriptor, currentValue *HistogramMetric) {
+	if currentValue == nil {
+		return
+	}
+
+	var entry histogramEntry
+	s.storeMutex.Lock()
+	if _, exists := s.store[metricDescriptor.Name]; !exists {
+		s.store[metricDescriptor.Name] = histogramEntry{}
+	}
+	entry = s.store[metricDescriptor.Name]
+	s.storeMutex.Unlock()
+
+	key := toHistogramKey(currentValue)
+	existing := entry[key]
+
+	if existing == nil {
+		level.Debug(s.logger).Log("msg", "Tracking new histogram", "fqName", currentValue.fqName, "key", key, "incoming_time", currentValue.reportTime)
+		entry[key] = &CollectedHistogram{histogram: currentValue, lastCollectedAt: time.Now()}
+		return
+	}
+
+	if existing.histogram.reportTime.Before(currentValue.reportTime) {
+		level.Debug(s.logger).Log("msg", "Incrementing existing histogram", "fqName", currentValue.fqName, "key", key, "last_reported_time", existing.histogram.reportTime, "incoming_time", currentValue.reportTime)
+		existing.histogram = mergeHistograms(existing.histogram, currentValue)
+		existing.lastCollectedAt = time.Now()
+		return
+	}
+
+	level.Debug(s.logger).Log("msg", "Ignoring old sample for histogram", "fqName", currentValue.fqName, "key", key, "last_reported_time", existing.histogram.reportTime, "incoming_time", currentValue.reportTime)
+}
+
+func toHistogramKey(hist *HistogramMetric) uint64 {
+	labels := make(map[string]string)
+	keysCopy := append([]string{}, hist.labelKeys...)
+	for i := range hist.labelKeys {
+		labels[hist.labelKeys[i]] = hist.labelValues[i]
+	}
+	sort.Strings(keysCopy)
+
+	var keyParts []string
+	for _, k := range keysCopy {
+		keyParts = append(keyParts, fmt.Sprintf("%s:%s", k, labels[k]))
+	}
+	hashText := fmt.Sprintf("%s|%s", hist.fqName, strings.Join(keyParts, "|"))
+	h := hashNew()
+	h = hashAdd(h, hashText)
+
+	return h
+}
+
+func mergeHistograms(existing *HistogramMetric, current *HistogramMetric) *HistogramMetric {
+	for key, value := range existing.buckets {
+		current.buckets[key] += value
+	}
+
+	// Calculate a new mean and overall count
+	mean := existing.dist.Mean
+	mean += current.dist.Mean
+	mean /= 2
+
+	var count uint64
```

Comment:
Shouldn't the count include the `existing.dist.Count`?

Reply:
This is safe because we copy the buckets from `existing` in to `current`, https://github.com/kgeckhart/stackdriver_exporter/blob/7d8ffbb7cfadc574ccbc094c94d0d09a8886cbfa/collectors/delta_distribution.go#L116-L118 at the end current is supposed to be `existing` merged with `current`.